File: apps/api/app/routers/v1/invitations.py
"""
Invitation management API endpoints.
"""

from typing import List, Optional
from fastapi import APIRouter, Depends, HTTPException, Query, status, BackgroundTasks
from sqlalchemy.orm import Session
from sqlalchemy import and_, or_

from app.database import get_db
from app.dependencies import get_current_user, require_org_admin
from app.models.invitation import (
    Invitation, InvitationStatus,
    InvitationCreate, InvitationUpdate, InvitationResponse,
    InvitationAcceptRequest, InvitationAcceptResponse,
    InvitationListResponse, BulkInvitationCreate, BulkInvitationResponse
)
from app.models.user import User
from app.services.invitation_service import InvitationService
from app.services.auth_service import AuthService
from app.config import settings


router = APIRouter(prefix="/v1/invitations", tags=["invitations"])


@router.post("/", response_model=InvitationResponse, status_code=status.HTTP_201_CREATED)
async def create_invitation(
    invitation_data: InvitationCreate,
    background_tasks: BackgroundTasks,
    current_user=Depends(require_org_admin),
    db: Session = Depends(get_db)
):
    """
    Create a new invitation for an organization (org admin only).
    """
    service = InvitationService(db)
    
    try:
        invitation = await service.create_invitation(
            invitation_data=invitation_data,
            invited_by=current_user,
            tenant_id=str(current_user.tenant_id)
        )
        
        return invitation
        
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )


@router.post("/bulk", response_model=BulkInvitationResponse, status_code=status.HTTP_201_CREATED)
async def create_bulk_invitations(
    bulk_data: BulkInvitationCreate,
    background_tasks: BackgroundTasks,
    current_user=Depends(require_org_admin),
    db: Session = Depends(get_db)
):
    """
    Create multiple invitations at once (org admin only).
    """
    service = InvitationService(db)
    
    result = await service.create_bulk_invitations(
        emails=bulk_data.emails,
        organization_id=bulk_data.organization_id,
        role=bulk_data.role,
        message=bulk_data.message,
        expires_in=bulk_data.expires_in,
        invited_by=current_user,
        tenant_id=str(current_user.tenant_id)
    )
    
    return BulkInvitationResponse(**result)


@router.get("/", response_model=InvitationListResponse)
async def list_invitations(
    organization_id: Optional[str] = None,
    status: Optional[InvitationStatus] = None,
    email: Optional[str] = None,
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=1000),
    current_user=Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    List invitations for organizations the user has access to.
    """
    query = db.query(Invitation).filter(
        Invitation.tenant_id == current_user.tenant_id
    )
    
    # Filter by organization if specified
    if organization_id:
        query = query.filter(Invitation.organization_id == organization_id)
    else:
        # Get user's organizations
        user_orgs = current_user.get_organizations()
        org_ids = [org.id for org in user_orgs]
        if org_ids:
            query = query.filter(Invitation.organization_id.in_(org_ids))
    
    # Filter by status
    if status:
        query = query.filter(Invitation.status == status.value)
    
    # Filter by email
    if email:
        query = query.filter(Invitation.email.ilike(f"%{email}%"))
    
    # Get total count
    total = query.count()
    
    # Get status counts
    pending_count = query.filter(Invitation.status == InvitationStatus.PENDING.value).count()
    accepted_count = query.filter(Invitation.status == InvitationStatus.ACCEPTED.value).count()
    expired_count = query.filter(Invitation.status == InvitationStatus.EXPIRED.value).count()
    
    # Get paginated results
    invitations = query.order_by(Invitation.created_at.desc()).offset(skip).limit(limit).all()
    
    # Convert to response models
    invitation_responses = []
    for inv in invitations:
        invitation_responses.append(InvitationResponse(
            id=str(inv.id),
            organization_id=str(inv.organization_id),
            email=inv.email,
            role=inv.role_name,
            status=inv.status,
            invited_by=str(inv.invited_by),
            message=inv.message,
            expires_at=inv.expires_at,
            created_at=inv.created_at,
            invite_url=inv.generate_invite_url(settings.APP_URL),
            email_sent=inv.email_sent
        ))
    
    return InvitationListResponse(
        invitations=invitation_responses,
        total=total,
        pending_count=pending_count,
        accepted_count=accepted_count,
        expired_count=expired_count
    )


@router.get("/{invitation_id}", response_model=InvitationResponse)
async def get_invitation(
    invitation_id: str,
    current_user=Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Get a specific invitation by ID.
    """
    invitation = db.query(Invitation).filter(
        and_(
            Invitation.id == invitation_id,
            Invitation.tenant_id == current_user.tenant_id
        )
    ).first()
    
    if not invitation:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Invitation not found"
        )
    
    return InvitationResponse(
        id=str(invitation.id),
        organization_id=str(invitation.organization_id),
        email=invitation.email,
        role=invitation.role_name,
        status=invitation.status,
        invited_by=str(invitation.invited_by),
        message=invitation.message,
        expires_at=invitation.expires_at,
        created_at=invitation.created_at,
        invite_url=invitation.generate_invite_url(settings.APP_URL),
        email_sent=invitation.email_sent
    )


@router.patch("/{invitation_id}", response_model=InvitationResponse)
async def update_invitation(
    invitation_id: str,
    update_data: InvitationUpdate,
    current_user=Depends(require_org_admin),
    db: Session = Depends(get_db)
):
    """
    Update a pending invitation (org admin only).
    """
    invitation = db.query(Invitation).filter(
        and_(
            Invitation.id == invitation_id,
            Invitation.tenant_id == current_user.tenant_id
        )
    ).first()
    
    if not invitation:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Invitation not found"
        )
    
    if invitation.status != InvitationStatus.PENDING.value:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Cannot update invitation with status: {invitation.status}"
        )
    
    # Update fields
    if update_data.role is not None:
        invitation.role_name = update_data.role
    
    if update_data.message is not None:
        invitation.message = update_data.message
    
    if update_data.expires_at is not None:
        invitation.expires_at = update_data.expires_at
    
    db.commit()
    db.refresh(invitation)
    
    return InvitationResponse(
        id=str(invitation.id),
        organization_id=str(invitation.organization_id),
        email=invitation.email,
        role=invitation.role_name,
        status=invitation.status,
        invited_by=str(invitation.invited_by),
        message=invitation.message,
        expires_at=invitation.expires_at,
        created_at=invitation.created_at,
        invite_url=invitation.generate_invite_url(settings.APP_URL),
        email_sent=invitation.email_sent
    )


@router.post("/{invitation_id}/resend", response_model=InvitationResponse)
async def resend_invitation(
    invitation_id: str,
    background_tasks: BackgroundTasks,
    current_user=Depends(require_org_admin),
    db: Session = Depends(get_db)
):
    """
    Resend an invitation email (org admin only).
    """
    service = InvitationService(db)
    
    try:
        invitation = await service.resend_invitation(
            invitation_id=invitation_id,
            resent_by=current_user
        )
        
        return invitation
        
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )


@router.delete("/{invitation_id}", status_code=status.HTTP_204_NO_CONTENT)
async def revoke_invitation(
    invitation_id: str,
    current_user=Depends(require_org_admin),
    db: Session = Depends(get_db)
):
    """
    Revoke a pending invitation (org admin only).
    """
    service = InvitationService(db)
    
    try:
        await service.revoke_invitation(
            invitation_id=invitation_id,
            revoked_by=current_user
        )
        
        return None
        
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )


@router.post("/accept", response_model=InvitationAcceptResponse)
async def accept_invitation(
    accept_data: InvitationAcceptRequest,
    db: Session = Depends(get_db)
):
    """
    Accept an invitation using the token.
    """
    service = InvitationService(db)
    auth_service = AuthService(db)
    
    try:
        # Get user if user_id provided
        user = None
        if accept_data.user_id:
            user = db.query(User).filter(
                User.id == accept_data.user_id
            ).first()
            
            if not user:
                raise ValueError("User not found")
        
        # Or create new user if registration data provided
        new_user_data = None
        if accept_data.password and accept_data.name and not user:
            password_hash = auth_service.hash_password(accept_data.password)
            new_user_data = {
                "name": accept_data.name,
                "password_hash": password_hash
            }
        
        # Accept invitation
        result = await service.accept_invitation(
            token=accept_data.token,
            user=user,
            new_user_data=new_user_data
        )
        
        return InvitationAcceptResponse(**result)
        
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )


@router.get("/validate/{token}")
async def validate_invitation_token(
    token: str,
    db: Session = Depends(get_db)
):
    """
    Validate an invitation token and get details.
    """
    invitation = db.query(Invitation).filter(
        Invitation.token == token
    ).first()
    
    if not invitation:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Invalid invitation token"
        )
    
    # Check if expired
    if invitation.is_expired:
        return {
            "valid": False,
            "reason": "Invitation has expired",
            "email": invitation.email,
            "organization_id": str(invitation.organization_id)
        }
    
    # Check if already accepted
    if invitation.status == InvitationStatus.ACCEPTED.value:
        return {
            "valid": False,
            "reason": "Invitation has already been accepted",
            "email": invitation.email,
            "organization_id": str(invitation.organization_id)
        }
    
    # Check if revoked
    if invitation.status == InvitationStatus.REVOKED.value:
        return {
            "valid": False,
            "reason": "Invitation has been revoked",
            "email": invitation.email,
            "organization_id": str(invitation.organization_id)
        }
    
    # Get organization details
    from app.models.organization import Organization
    org = db.query(Organization).filter(
        Organization.id == invitation.organization_id
    ).first()
    
    return {
        "valid": True,
        "email": invitation.email,
        "organization_id": str(invitation.organization_id),
        "organization_name": org.name if org else None,
        "role": invitation.role_name,
        "expires_at": invitation.expires_at.isoformat(),
        "message": invitation.message
    }


@router.post("/cleanup")
async def cleanup_expired_invitations(
    current_user=Depends(require_org_admin),
    db: Session = Depends(get_db)
):
    """
    Clean up expired invitations (org admin only).
    """
    service = InvitationService(db)
    
    count = service.cleanup_expired_invitations()
    
    return {
        "message": f"Marked {count} invitations as expired",
        "count": count
    }

File: packages/typescript-sdk/src/invitations.ts
/**
 * Invitations module for the Janua TypeScript SDK
 *
 * Provides organization invitation management capabilities.
 */

import type { HttpClient } from './http-client';
import type { UUID, ISODateString } from './types';
import { ValidationError } from './errors';
import { ValidationUtils } from './utils';

// Invitation Types

export enum InvitationStatus {
  PENDING = 'pending',
  ACCEPTED = 'accepted',
  EXPIRED = 'expired',
  REVOKED = 'revoked'
}

export interface InvitationCreate {
  organization_id: UUID;
  email: string;
  role?: string; // 'owner' | 'admin' | 'member' | 'viewer'
  message?: string;
  expires_in?: number; // days (1-30)
}

export interface InvitationUpdate {
  role?: string;
  message?: string;
  expires_at?: ISODateString;
}

export interface Invitation {
  id: UUID;
  organization_id: UUID;
  email: string;
  role: string;
  status: InvitationStatus;
  invited_by: UUID;
  message?: string;
  expires_at: ISODateString;
  created_at: ISODateString;
  invite_url: string;
  email_sent: boolean;
}

export interface InvitationListParams {
  organization_id?: UUID;
  status?: InvitationStatus;
  email?: string;
  skip?: number;
  limit?: number;
}

export interface InvitationListResponse {
  invitations: Invitation[];
  total: number;
  pending_count: number;
  accepted_count: number;
  expired_count: number;
}

export interface InvitationAcceptRequest {
  token: string;
  user_id?: UUID; // If user already exists
  // For new user registration
  name?: string;
  password?: string;
}

export interface InvitationAcceptResponse {
  success: boolean;
  user_id: UUID;
  organization_id: UUID;
  message: string;
  is_new_user: boolean;
}

export interface InvitationValidateResponse {
  valid: boolean;
  reason?: string;
  email?: string;
  organization_id?: UUID;
  organization_name?: string;
  role?: string;
  expires_at?: ISODateString;
  message?: string;
}

export interface BulkInvitationCreate {
  organization_id: UUID;
  emails: string[];
  role?: string;
  message?: string;
  expires_in?: number; // days
}

export interface BulkInvitationResponse {
  successful: Invitation[];
  failed: Array<{ email: string; error: string }>;
  total_sent: number;
  total_failed: number;
}

/**
 * Invitation management operations
 */
export class Invitations {
  constructor(private http: HttpClient) {}

  /**
   * Create a new invitation
   *
   * @param invitation - Invitation details
   * @returns Created invitation
   */
  async createInvitation(invitation: InvitationCreate): Promise<Invitation> {
    // Validate organization ID
    if (!ValidationUtils.isValidUuid(invitation.organization_id)) {
      throw new ValidationError('Invalid organization ID format');
    }

    // Validate email
    if (!ValidationUtils.isValidEmail(invitation.email)) {
      throw new ValidationError('Invalid email format');
    }

    // Validate role
    if (invitation.role && !['owner', 'admin', 'member', 'viewer'].includes(invitation.role)) {
      throw new ValidationError('Invalid role. Must be owner, admin, member, or viewer');
    }

    // Validate message length
    if (invitation.message && invitation.message.length > 500) {
      throw new ValidationError('Message must be 500 characters or less');
    }

    // Validate expires_in
    if (invitation.expires_in !== undefined) {
      if (invitation.expires_in < 1 || invitation.expires_in > 30) {
        throw new ValidationError('Expiration must be between 1 and 30 days');
      }
    }

    const response = await this.http.post<Invitation>('/api/v1/invitations', invitation);
    return response.data;
  }

  /**
   * Create multiple invitations at once
   *
   * @param bulkInvitation - Bulk invitation details
   * @returns Result with successful and failed invitations
   */
  async createBulkInvitations(bulkInvitation: BulkInvitationCreate): Promise<BulkInvitationResponse> {
    // Validate organization ID
    if (!ValidationUtils.isValidUuid(bulkInvitation.organization_id)) {
      throw new ValidationError('Invalid organization ID format');
    }

    // Validate emails
    if (!bulkInvitation.emails || bulkInvitation.emails.length === 0) {
      throw new ValidationError('At least one email is required');
    }

    if (bulkInvitation.emails.length > 100) {
      throw new ValidationError('Maximum 100 emails allowed per bulk invitation');
    }

    for (const email of bulkInvitation.emails) {
      if (!ValidationUtils.isValidEmail(email)) {
        throw new ValidationError(`Invalid email format: ${email}`);
      }
    }

    // Validate role
    if (bulkInvitation.role && !['owner', 'admin', 'member', 'viewer'].includes(bulkInvitation.role)) {
      throw new ValidationError('Invalid role. Must be owner, admin, member, or viewer');
    }

    // Validate message length
    if (bulkInvitation.message && bulkInvitation.message.length > 500) {
      throw new ValidationError('Message must be 500 characters or less');
    }

    // Validate expires_in
    if (bulkInvitation.expires_in !== undefined) {
      if (bulkInvitation.expires_in < 1 || bulkInvitation.expires_in > 30) {
        throw new ValidationError('Expiration must be between 1 and 30 days');
      }
    }

    const response = await this.http.post<BulkInvitationResponse>(
      '/api/v1/invitations/bulk',
      bulkInvitation
    );
    return response.data;
  }

  /**
   * List invitations with optional filtering
   *
   * @param params - Filter and pagination parameters
   * @returns Paginated invitation list with counts
   */
  async listInvitations(params?: InvitationListParams): Promise<InvitationListResponse> {
    // Validate organization ID if provided
    if (params?.organization_id && !ValidationUtils.isValidUuid(params.organization_id)) {
      throw new ValidationError('Invalid organization ID format');
    }

    // Validate status if provided
    if (params?.status && !Object.values(InvitationStatus).includes(params.status)) {
      throw new ValidationError('Invalid invitation status');
    }

    // Validate pagination
    if (params?.skip !== undefined && params.skip < 0) {
      throw new ValidationError('Skip must be non-negative');
    }

    if (params?.limit !== undefined) {
      if (params.limit < 1 || params.limit > 1000) {
        throw new ValidationError('Limit must be between 1 and 1000');
      }
    }

    // Build query string
    const queryParams = new URLSearchParams();
    if (params?.organization_id) queryParams.append('organization_id', params.organization_id);
    if (params?.status) queryParams.append('status', params.status);
    if (params?.email) queryParams.append('email', params.email);
    if (params?.skip !== undefined) queryParams.append('skip', params.skip.toString());
    if (params?.limit !== undefined) queryParams.append('limit', params.limit.toString());

    const queryString = queryParams.toString();
    const url = queryString ? `/api/v1/invitations?${queryString}` : '/api/v1/invitations';

    const response = await this.http.get<InvitationListResponse>(url);
    return response.data;
  }

  /**
   * Get a specific invitation by ID
   *
   * @param invitationId - Invitation ID
   * @returns Invitation details
   */
  async getInvitation(invitationId: UUID): Promise<Invitation> {
    if (!ValidationUtils.isValidUuid(invitationId)) {
      throw new ValidationError('Invalid invitation ID format');
    }

    const response = await this.http.get<Invitation>(`/api/v1/invitations/${invitationId}`);
    return response.data;
  }

  /**
   * Update a pending invitation
   *
   * @param invitationId - Invitation ID
   * @param update - Fields to update
   * @returns Updated invitation
   */
  async updateInvitation(invitationId: UUID, update: InvitationUpdate): Promise<Invitation> {
    if (!ValidationUtils.isValidUuid(invitationId)) {
      throw new ValidationError('Invalid invitation ID format');
    }

    // Validate role if provided
    if (update.role && !['owner', 'admin', 'member', 'viewer'].includes(update.role)) {
      throw new ValidationError('Invalid role. Must be owner, admin, member, or viewer');
    }

    // Validate message length if provided
    if (update.message && update.message.length > 500) {
      throw new ValidationError('Message must be 500 characters or less');
    }

    const response = await this.http.patch<Invitation>(
      `/api/v1/invitations/${invitationId}`,
      update
    );
    return response.data;
  }

  /**
   * Resend an invitation email
   *
   * @param invitationId - Invitation ID
   * @returns Updated invitation
   */
  async resendInvitation(invitationId: UUID): Promise<Invitation> {
    if (!ValidationUtils.isValidUuid(invitationId)) {
      throw new ValidationError('Invalid invitation ID format');
    }

    const response = await this.http.post<Invitation>(
      `/api/v1/invitations/${invitationId}/resend`,
      {}
    );
    return response.data;
  }

  /**
   * Revoke (cancel) a pending invitation
   *
   * @param invitationId - Invitation ID
   */
  async revokeInvitation(invitationId: UUID): Promise<void> {
    if (!ValidationUtils.isValidUuid(invitationId)) {
      throw new ValidationError('Invalid invitation ID format');
    }

    await this.http.delete(`/api/v1/invitations/${invitationId}`);
  }

  /**
   * Accept an invitation using the token
   *
   * @param request - Accept request with token and optional user info
   * @returns Acceptance result
   */
  async acceptInvitation(request: InvitationAcceptRequest): Promise<InvitationAcceptResponse> {
    if (!request.token || request.token.trim().length === 0) {
      throw new ValidationError('Invitation token is required');
    }

    // Validate user_id if provided
    if (request.user_id && !ValidationUtils.isValidUuid(request.user_id)) {
      throw new ValidationError('Invalid user ID format');
    }

    // Validate new user registration if provided
    if (request.password && !request.user_id) {
      if (!request.name || request.name.trim().length === 0) {
        throw new ValidationError('Name is required for new user registration');
      }

      if (request.name.length > 255) {
        throw new ValidationError('Name must be 255 characters or less');
      }

      const passwordValidation = ValidationUtils.validatePassword(request.password);
      if (!passwordValidation.isValid) {
        throw new ValidationError(
          'Password validation failed',
          passwordValidation.errors.map(err => ({ field: 'password', message: err }))
        );
      }
    }

    const response = await this.http.post<InvitationAcceptResponse>(
      '/api/v1/invitations/accept',
      request
    );
    return response.data;
  }

  /**
   * Validate an invitation token
   *
   * @param token - Invitation token
   * @returns Validation result with invitation details
   */
  async validateToken(token: string): Promise<InvitationValidateResponse> {
    if (!token || token.trim().length === 0) {
      throw new ValidationError('Invitation token is required');
    }

    const response = await this.http.get<InvitationValidateResponse>(
      `/api/v1/invitations/validate/${token}`
    );
    return response.data;
  }

  /**
   * Clean up expired invitations (admin only)
   *
   * @returns Count of invitations marked as expired
   */
  async cleanupExpiredInvitations(): Promise<{ message: string; count: number }> {
    const response = await this.http.post<{ message: string; count: number }>(
      '/api/v1/invitations/cleanup',
      {}
    );
    return response.data;
  }
}
